packages/wdio-cli/src/launcher.js

```diff
@@ -53,6 +50,46 @@ class Launcher {
         this.runnerFailed = 0
     }
 
+    /**
+     * calculates the worker count based on dataprovider injected for each spec file
+     * @return {Number} the worker count required for each capability
+     */
+    calculateWorkerCountForCapability(c) {
+        let specs = this.getSpecsWithDataProviderAndRetries(c.specs, c.exclude)
+        return specs.length
+    }
+
+    /**
+     * Combinate test data from data provider and retries with specs
+     */
+    getSpecsWithDataProviderAndRetries(capSpecs, capExclude) {
+
+        /**
+         * avoid retries in watch mode
+         */
+        let config = this.configParser.getConfig()
+        const specFileRetries = config.watch ? 0 : config.specFileRetries
+        let specs = []
+        const dataProvidersMap = this.configParser.getDataProviders()
+
+        this.configParser.getSpecs(capSpecs, capExclude).forEach((specFile) => {
+            if (dataProvidersMap != null) {
```

Comment:
```suggestion
            if (dataProvidersMap && dataProvidersMap[specFile]) {
```

with that we can removed ` if (typeof dataProvider === 'undefined') {` 2 lines below

Reply:
I have updated the recommended changes.